src/mlpack/methods/linear_svm/linear_svm_function_impl.hpp

```diff
@@ -0,0 +1,352 @@
+  weights *= 0.005;
+}
+
+/**
+ * This is equivalent to applying the indicator function to the training
+ * labels. The output is in the form of a matrix, which leads to simpler
+ * calculations in the Evaluate() and Gradient() methods.
+ */
+template <typename MatType>
+void LinearSVMFunction<MatType>::GetGroundTruthMatrix(
+    const arma::Row<size_t>& labels,
+    arma::sp_mat& groundTruth)
+{
+  // Calculate the ground truth matrix according to the labels passed. The
+  // ground truth matrix is a matrix of dimensions 'numClasses * numExamples',
+  // where each column contains a single entry of '1', marking the label
+  // corresponding to that example.
+
+  // Row pointers and column pointers corresponding to the entries.
+  arma::uvec rowPointers(labels.n_elem);
+  arma::uvec colPointers(labels.n_elem + 1);
+
+  // Row pointers are the labels of the examples, and column pointers are the
+  // number of cumulative entries made uptil that column.
+  for (size_t i = 0; i < labels.n_elem; i++)
+  {
+    rowPointers(i) = labels(i);
+    colPointers(i + 1) = i + 1;
+  }
+
+  // All entries are '1'.
+  arma::vec values;
+  values.ones(labels.n_elem);
+
+  // Calculate the matrix.
+  groundTruth = arma::sp_mat(rowPointers, colPointers, values, numClasses,
+                             labels.n_elem);
+}
+
+/**
+ * Shuffle the data.
+ */
+template <typename MatType>
+void LinearSVMFunction<MatType>::Shuffle()
+{
+  // Determine new ordering.
+  arma::uvec ordering = arma::shuffle(arma::linspace<arma::uvec>(0,
+      dataset.n_cols - 1, dataset.n_cols));
+
+  // Re-sort data.
+  arma::mat newData = dataset.cols(ordering);
+  math::ClearAlias(dataset);
+  dataset = std::move(newData);
+
+  // Assemble data for batch constructor.  We need reverse orderings though...
+  arma::uvec reverseOrdering(ordering.n_elem);
+  for (size_t i = 0; i < ordering.n_elem; ++i)
+    reverseOrdering[ordering[i]] = i;
+
+  arma::umat newLocations(2, groundTruth.n_nonzero);
+  arma::vec values(groundTruth.n_nonzero);
+  arma::sp_mat::const_iterator it = groundTruth.begin();
+  size_t loc = 0;
+  while (it != groundTruth.end())
+  {
+    newLocations(0, loc) = reverseOrdering(it.col());
+    newLocations(1, loc) = it.row();
+    values(loc) = (*it);
+
+    ++it;
+    ++loc;
+  }
+
+  groundTruth = arma::sp_mat(newLocations, values, groundTruth.n_rows,
+                             groundTruth.n_cols);
+}
+
+template <typename MatType>
+double LinearSVMFunction<MatType>::Evaluate(
+    const arma::mat& parameters)
+{
+  // The objective function is the hinge loss function and it is
+  // calculated over all the training examples.
+
+  // Calculate the loss and regularization terms.
+  double loss, regularization, cost;
+
+  arma::mat scores = parameters * dataset;
+  arma::sp_mat correctScores = scores % groundTruth;
+  arma::mat margin = scores
+      - arma::repmat(arma::ones(numClasses).t() * correctScores, numClasses, 1)
+      + 1 - groundTruth;
```

Comment:
Hmm, I'm not easily following this (that doesn't mean it's wrong, just that I don't 100% understand yet).  Working from eq. (2) in https://www.elen.ucl.ac.be/Proceedings/esann/esannpdf/es1999-461.pdf, I found that the optimization is equivalent to

```
min 0.5 Σ_m (w_m * w_m) + C Σ_i Σ_m max(0, 2 + (w_m x_i + b_m) - (w_{y_i} x_i + b_{y_i}))
```

so there is a `2` instead of a `1`.  This matches your derivation provided in an earlier comment but I am not sure what the value of Δ is in what you gave there.  Can you clarify a little bit please?  I may have done something wrong.  Also, if you could work that explanation into the comment here, it would make it a lot easier for future maintainers. :+1:  It's currently a little hard to see how to go from equation (2) given in that paper to this formulation here, which uses sparse matrices and sparse linear algebra.

Thanks!

Reply:
Thanks for pointing this, turns out, I've been using `1` as the value of the hyperparameter `Δ` while evaluating the Loss Function. You were right in deriving this. 
```
min 0.5 Σ_m (w_m * w_m) + C Σ_i Σ_m max(0, Δ + (w_m x_i + b_m) - (w_{y_i} x_i + b_{y_i}))
```


However, since I used Δ = 1, my loss function over there was 
```
min 0.5 Σ_m (w_m * w_m) + C Σ_i Σ_m max(0, 1 + (w_m x_i + b_m) - (w_{y_i} x_i + b_{y_i}))
```

The way I'm programmed it over there is: 
```
  // this simply evaluates the score of each class 
  arma::mat scores = parameters * dataset;

  // for each column, only the row(which indicates the correct label)
  // holds the value of the score.
  arma::sp_mat correctScores = scores % groundTruth;

  // this step does alot of things in a single line. 
  // first 
  // arma::repmat(arma::ones(numClasses).t() * correctScores, numClasses, 1)
  // this gives us a matrix where each row of any specific column is having the 
  // value of correct class score for that column.
  // Now,
  arma::mat margin = scores - 
  		arma::repmat(arma::ones(numClasses).t() * 
  		correctScores, numClasses, 1) + 1 - groundTruth;

  // Ignore the '1's here, the Δ is assumed to be 1 here (as of this review).
  // Now
  // the operation above now slightly corresponds to 		
  // Σ_i Σ_m max(0, 1 + (w_m x_i + b_m) - (w_{y_i} x_i + b_{y_i}))
  //
  // The reason I subtracted groundTruth in the end is to make the 
  // margin of the correct label in each column to be `0`, which is because 
  //  Σ_m is taken over every m != y_i. (Equation 2 of the given paper)
  // whereas I added `1` ( Δ Hyperparamtere ) to every element of the margin.
```

I made a small code that shows this operation(just by printing each matrix after every step, nothing more than that) and you can check it out here: https://github.com/AYESDIE/random-code/tree/master/mlpack-svmtest

I'll address these changes pretty soon.

Reply:
Also, I used http://cs231n.github.io/linear-classify/#svm for reference.

Reply:
Ah, ok.  I think it is fine to go with Δ = 2, since that keeps it closer to the original formulation.  The link you gave (to the cs231n site) is correct that Δ can take different values and the result will still be generally the same, but there is no direct mapping between Δ and λ; i.e., we aren't guaranteed that we can optimize the same objective function with Δ = 1 and Δ = 2.